cpp/test/mg/pca.cu

```diff
@@ -0,0 +1,214 @@
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include <common/cudart_utils.h>
+#include <test_utils.h>
+#include <gtest/gtest.h>
+#include <linalg/cublas_wrappers.h>
+#include <cuda_utils.cuh>
+#include <cuml/decomposition/pca_mg.hpp>
+#include <opg/linalg/gemm.hpp>
+#include <opg/matrix/matrix_utils.hpp>
+#include <matrix/matrix.cuh>
+#include "test_opg_utils.h"
+
+namespace MLCommon {
+namespace Test {
+namespace opg {
+
+struct PCAOpgParams {
+  int M;
+  int N;
+  int N_components;
+  ML::mg_solver algorithm;
+  std::vector<int> partSizes;
+  std::vector<int> ranksOwners;
+  Matrix::Layout layout;
+  unsigned long long int seed;
+};
+
+template <typename T>
+class PCAOpgTest : public testing::TestWithParam<PCAOpgParams> {
+ public:
+  void SetUp() {
+    params = GetParam();
+    handle = new ML::cumlHandle();
+    ML::initialize_mpi_comms(*handle, MPI_COMM_WORLD);
+
+    // Prepare resource
+    const ML::cumlHandle_impl& h = handle->getImpl();
+    const cumlCommunicator& comm = h.getCommunicator();
+    stream = h.getStream();
+    const std::shared_ptr<deviceAllocator> allocator = h.getDeviceAllocator();
+    cublasHandle_t cublasHandle = h.getCublasHandle();
+
+    myRank = comm.getRank();
+    totalRanks = comm.getSize();
+    Random::Rng r(params.seed + myRank);
+
+    CUBLAS_CHECK(cublasSetStream(cublasHandle, stream));
+
+    if (myRank == 0) {
+      std::cout << "Testing PCA of " << params.M << " x " << params.N
+                << " matrix" << std::endl;
+    }
+
+    // Prepare X matrix
+    std::vector<Matrix::RankSizePair*> totalPartsToRanks;
+    for (int i = 0; i < params.partSizes.size(); i++) {
+      Matrix::RankSizePair* rspt = new Matrix::RankSizePair(
+        params.ranksOwners[i] % totalRanks, params.partSizes[i]);
+      totalPartsToRanks.push_back(rspt);
+    }
+    Matrix::PartDescriptor desc(params.M, params.N, totalPartsToRanks,
+                                comm.getRank(), params.layout);
+    std::vector<Matrix::Data<T>*> inParts;
+    Matrix::opg::allocate(h, inParts, desc, myRank, stream);
+    Matrix::opg::randomize(h, r, inParts, desc, myRank, stream, T(10.0),
+                           T(20.0));
+    h.waitOnUserStream();
+
+    prmsPCA.n_rows = params.M;
+    prmsPCA.n_cols = params.N;
+    prmsPCA.n_components = params.N_components;
+    prmsPCA.whiten = false;
+    prmsPCA.n_iterations = 100;
+    prmsPCA.tol = 0.01;
+    prmsPCA.algorithm = params.algorithm;
+
+    components = (T*)allocator->allocate(
+      prmsPCA.n_components * prmsPCA.n_cols * sizeof(T), stream);
+
+    explained_var =
```

Comment:
We should have access to device_buffer at this layer. Would be good to use those so reduce the potential for leaks.

Reply:
Using the allocator as we are printing these arrays as debug options below

Reply:
That doesn’t have any effect on our ability to use allocator vs device_buffer. You would just need to call `buffer_ptr.data()`. We really should be using the device_buffer everywhere since it is RAII and not using trhe allocator directly anymore.